molfeat/trans/pretrained/base.py

```diff
@@ -20,28 +20,39 @@ class PretrainedMolTransformer(MoleculeTransformer):
     r"""
     Transformer based on pretrained featurizer
 
+    !!! note
+        * When implementing a subclass of this class, you need to define the `_embed` and optionally the `_convert` methods.
+        * If your model is an instance of PretrainedModel that handles loading of the model from a store, then
+        you can decide whether you want to preload the true underlying model by loading it from the store.
+        You will be in charge of handling the logic of when you need to call preload, and when you don't.
+        Note however that by default preloading is only attempted when the featurizer is still an instance of PretrainedModel.
```

Comment:
This is still not entirely clear to me. Let me try summarize it in my own words to see if I understood it: 

----

The `PretrainedMolTransformer` is the best class to subclass if you want to add your own pre-trained featurizer. The class supports caching out-of-the-box. To remain flexible in which pre-trained featurizers we support, we do not impose any constraints on the type or interface of the pretrained model and rather leave it up to the user to override any method of the `PretrainedMolTransformer` as needed. Any subclass only _has to_ override the  `_embed` method. 

To simplify the implementation of new pre-trained featurizers, however, we do supply users with the `_convert` method and provide the `PretrainedModel` interface. 
- The `_convert()` method transforms a set of smiles to the input format expected by the pre-trained featurizers (e.g. a set of PyG graphs for a GNN). 
- If the `featurizer` is a subclass of the `PretrainedModel` interface, we will be able to automatically load it if you specify `preload=True`. 

---

Some questions remain: 
- Why do you mention "loading from the _store_" for the `PretrainedModel`? Doesn't have to be from the store right? Seems that is what the `PretrainedStoreModel` is for. 
- Shouldn't the `load()` method of the `PretrainedModel` and `PretrainedStoreModel` have a default implementation or be an abstract method?
- Should `_embed()` (and `_convert()`) be public facing? Then it's at least clearer to me personally what I, as a user subclassing this, am supposed to touch and what not from the method name alone. 
- Could you name some examples in which using `PretrainedModel`, `_embed()` and `_convert()` is not enough? I cannot think of any! If none, why not enforce the type of `featurizer` in this class to be a `PretrainedModel`? I would expect that to stimulate a more uniform code base.

Reply:
If I understood this well, I'll change the tutorial to use the `PretrainedModel` interface.

Reply:
> This is still not entirely clear to me. Let me try summarize it in my own words to see if I understood it:
> 
> The `PretrainedMolTransformer` is the best class to subclass if you want to add your own pre-trained featurizer. The class supports caching out-of-the-box. To remain flexible in which pre-trained featurizers we support, we do not impose any constraints on the type or interface of the pretrained model and rather leave it up to the user to override any method of the `PretrainedMolTransformer` as needed. Any subclass only _has to_ override the `_embed` method.

yes

> To simplify the implementation of new pre-trained featurizers, however, we do supply users with the `_convert` method and provide the `PretrainedModel` interface.

The `PretrainedModel` interfaces handles how pretrained model are loaded, as loading a pretrained model often involves more than loading the weights and handling those additional processes in the `PretrainedMolTransformer` is not practical. We likely would like to prevent model loading outside of our store.  

> 
> * The `_convert()` method transforms a set of smiles to the input format expected by the pre-trained featurizers (e.g. a set of PyG graphs for a GNN).

> * If the `featurizer` is a subclass of the `PretrainedModel` interface, we will be able to automatically load it if you specify `preload=True`. 

Loading a large model take resources, and we would like to avoid assigning resources (RAM, DISK, CPUs) unless required. However, loading the featurizer in memory from the start might be the preferred options in some rare cases. 

Regardless of whether you set `preload` to True or not,  your model will be loaded before embedding. `Preload`  helps you control whether you want to load the model on object initialization or not. 

- If your model is not a `PretrainedModel` it does not impact anything
- You are always free to handle how you load your model in `PretrainedMolTransformer` 
- Implementing `_preload` is not required. This is purely an optimization goal.

> 
> Some questions remain:
> 
> * Why do you mention "loading from the _store_" for the `PretrainedModel`? Doesn't have to be from the store right? Seems that is what the `PretrainedStoreModel` is for.

Changed, it's either the store or through any complex mechanism. 

> * Shouldn't the `load()` method of the `PretrainedModel` and `PretrainedStoreModel` have a default implementation or be an abstract method?

I changed `load` to abstractmethod in `PretrainedModel`.  `Raise Implementation error` is the right call for  `PretrainedStoreModel`


> * Should `_embed()` (and `_convert()`) be public facing? Then it's at least clearer to me personally what I, as a user subclassing this, am supposed to touch and what not from the method name alone.

They are not supposed to be called by end user externally but defined to extend functionality so they are `protected` methods (and not private methods), which is the intended access level.

> * Could you name some examples in which using `PretrainedModel`, `_embed()` and `_convert()` is not enough? I cannot think of any! If none, why not enforce the type of `featurizer` in this class to be a `PretrainedModel`? I would expect that to stimulate a more uniform code base.

```bash
pip install astrochem_embedding
```
```python

import joblib
import fsspec 
import torch
import datamol as dm
from astrochem_embedding import VICGAE
from molfeat.trans.pretrained import PretrainedMolTransformer

class MyLazyAstroChemFeaturizer(PretrainedMolTransformer):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        # this user does not care about the molfeat store or
        # does not want to define a new class just to run a single line below or
        # does not care about any memory optimization mechanism (preload)
        self.featurizer = VICGAE.from_pretrained()
    
    def _embed(self, smiles, **kwargs):
        return [self.featurizer.embed_smiles(x) for x in smiles]

transformer = MyLazyAstroChemFeaturizer(dtype=torch.float)
transformer(dm.freesolv()["smiles"][:10])
```

Reply:
Mostly clear and in line with what I said I think!

```
# this user does not care about the molfeat store or
# does not want to define a new class just to run a single line below or
# does not care about any memory optimization mechanism (preload)
```

You didn't answer my question here, I wondered in which case it's _not_ sufficient. In this case it's still sufficient (you don't even need it all).

Reply:
Sorry, got it wrong. Was replying to forcing people to implement a subclass of `PretrainedModel`. 

I think most use cases are covered by the above logic.  Only case this could be inefficient is if embedding performed by your model cannot be batched. Then you might want to redefine some additional logics in `transform` and `_transform` to fully leverage parallelization when needed. 

Unlikely to be the case for most pretrained models. And it's always possible to use `MoleculeTransformer`, with a `calculator` instead.